ResponseHandler_NotifyTest.normalResponse_whenBackupMissing_thenEventuallySuccess

```
java.lang.AssertionError: expected null, but was:<Invocation{op=com.hazelcast.spi.impl.operationservice.impl.DummyBackupAwareOperation{serviceName='null', identityHash=582573722, partitionId=1, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}, tryCount=0, tryPauseMillis=0, invokeCount=0, callTimeoutMillis=20000, firstInvocationTimeMs=1464920582458, firstInvocationTime='2016-06-03 05:23:02.458', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[127.0.0.1]:15011, pendingResponse={foo}, backupsAcksExpected=1, backupsAcksReceived=0, connection=null}>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotNull(Assert.java:755)
	at org.junit.Assert.assertNull(Assert.java:737)
	at org.junit.Assert.assertNull(Assert.java:747)
	at com.hazelcast.spi.impl.operationservice.impl.ResponseHandler_NotifyTest.normalResponse_whenBackupMissing_thenEventuallySuccess(ResponseHandler_NotifyTest.java:117)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK8/com.hazelcast$hazelcast/823/testReport/junit/com.hazelcast.spi.impl.operationservice.impl/ResponseHandler_NotifyTest/normalResponse_whenBackupMissing_thenEventuallySuccess/

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/ResponseHandler_NotifyTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/ResponseHandler_NotifyTest.java
@@ -4,6 +4,7 @@ import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.OperationTimeoutException;
 import com.hazelcast.spi.Operation;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.ExpectedRuntimeException;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -69,7 +70,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
         responseHandler.notifyNormalResponse(callId, value, 0, null);
 
         assertEquals(value, invocation.future.join());
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
     }
 
     @Test
@@ -81,7 +82,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
 
         responseHandler.notifyNormalResponse(callId, "foo", 0, null);
 
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
     }
 
     @Test
@@ -96,7 +97,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
 
         Object value = "foo";
         responseHandler.notifyNormalResponse(callId, value, 1, null);
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
 
         assertEquals(value, invocation.future.join());
     }
@@ -108,13 +109,13 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
         Invocation invocation = newInvocation();
 
         invocationRegistry.register(invocation);
-        long callId = invocation.op.getCallId();
+        final long callId = invocation.op.getCallId();
 
         String result = "foo";
         responseHandler.notifyNormalResponse(callId, result, 1, null);
 
         assertEquals(result, invocation.future.get(1, TimeUnit.MINUTES));
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
     }
 
     @Test
@@ -135,7 +136,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
         assertSame(invocation, invocationRegistry.get(callId));
 
         responseHandler.notifyBackupComplete(callId);
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
 
         assertEquals(value, invocation.future.join());
     }
@@ -148,7 +149,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
         invocationRegistry.deregister(invocation);
 
         responseHandler.notifyBackupComplete(callId);
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
     }
 
     // ==================== errorResponse ======================
@@ -167,7 +168,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
         } catch (ExpectedRuntimeException expected) {
         }
 
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
     }
 
     @Test
@@ -179,7 +180,7 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
 
         responseHandler.notifyErrorResponse(callId, new ExpectedRuntimeException(), null);
 
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
     }
 
     // ==================== timeoutResponse =====================
@@ -198,6 +199,15 @@ public class ResponseHandler_NotifyTest extends HazelcastTestSupport {
         } catch (OperationTimeoutException expected) {
         }
 
-        assertNull(invocationRegistry.get(callId));
+        assertInvocationDeregisteredEventually(callId);
+    }
+
+    private void assertInvocationDeregisteredEventually(final long callId) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertNull(invocationRegistry.get(callId));
+            }
+        });
     }
 }

